### PR TITLE
Don't block render with process.nextTick polyfill.

### PR DESCRIPTION
**Pull request by mikeal**
If you slam the browserify process.nextTicket polyfill it appears to block UI rendering, scrolling, etc.

These methods ensure that no more than one callback comes off the event queue at a time. I'm running it locally and it completely fixed an issue I had. 

I tried to match the coding style in the file but I'm not 100% sure I nailed it.

### CODE REVIEW

**Comment by vweevers** [comment]
This will increase latency though, correct? Because the browserify version avoids setTimeout on recursive ticks.

Have you tried including a `setImmediate` polyfill in your app?

**Comment by calvinmetcalf** [comment]
yeah @mikeal the setImmediate polyfil will cause this to pick it up and will have behavior similar to what you did here but with less latency

**Comment by mikeal** [comment]
I'm pretty sure it would hit the same problem with a setImmediate polyfill (which operates similarly to the nextTick polyfill). If you do 10 operations in a single tick they will all come out of nextTick/setImmediate at the same time in one batch without yielding to the rendered or anything else in the event queue.

Not sure I buy the latency argument. 1) this is memdown, even with this change everything is an order of magnitude faster than a leveldown implementation that touches the disc and 2) the only "latency" it ads is allowing other events and rendering to happen between database operations, which is actually the entire point ;)

That said, if this is in conflict with other people's use cases I can just publish this as a fork for my own use.

**Comment by mikeal** [comment]
A little more on where I'm seeing this as an issue.

I'm using `hyperlog` in a p2p app and when the initial replicator is setup and pulls in a new database it processes as many records as exist in a chunk of data (between 10 and 30) and it hits the local levelup instance for that many records. When those record lookups return without any data it then has to run a signature verification on every record. Because all 10 - 30 records come out in the same tick rendering/scrolling/usability completely stop for 6-10 full seconds while the signature verification runs.

This goes away entirely when using any other leveldown (because they touch disc and don't all come out in the same tick) and it also goes away with this patch.

**Comment by calvinmetcalf** [comment]
if you used something which checked the signatures async (e.g. anything that uses browser crypto) you would totally side step the problem, the same if you add this queue to the signature verification function

**Comment by mikeal** [comment]
@calvinmetcalf I could, it just seems odd to expect the crypto (traditionally sync) to deal with this rather than the database (which is traditionally async). It's also a potential differentiator between memdown and all other leveldown implementations, but maybe I'm overthinking it.

**Comment by calvinmetcalf** [comment]
the issue isn't that the database isn't async it's that the database callbacks are all called in a row without the event loop twirling between them, which usually doesn't happen with a database but there is no reason it couldn't especially if stuff was bundled

> it just seems odd to expect the crypto (traditionally sync) to deal with this 

in the browser built in crypto is async so switching to a built in async crypto would probably help you regardless of anything else

**Comment by nolanlawson** [comment]
Just piping up to say I think it's important to test this in multiple browsers, since Edge and IE are the only ones with `setImmediate` and that will probably always be the case since we never got buy-in from other vendors. And `setImmediate()` in the MS browsers isn't doing quite the same thing that it does in Node (e.g. it helps unblock input, so @mikeal if you run your example in Edge you'll probably find that scrolling is fine, especially non-keyboard scrolling).

So basically I find the entire `global.setImmediate || ...` usage here suspect and would imagine it's better to just use `setImmediate()` in Node and some other entirely different thing for browsers. What that should be is debatable, though, and I'd defer to @calvinmetcalf since he's the expert. :)

**Comment by nolanlawson** [comment]
@mikeal Can you provide a demo page to show this 30-second delay? I'd like to see if this is really something that ought to be solved by memdown or ought to be solved in userland by using a bulk/iterator API instead of a single get/set API.

Looking at browserify's `process.nextTick()` shim, it seems to essentially be a `setTimeout(fn, 0)` so I'm not following how it would be different from the proposal in this PR.

**Comment by calvinmetcalf** [comment]
@nolanlawson *nested* process.nextTick or calls a while one is in progress will be sync.  Since what @mikeal is doing is requesting a bunch of records, in memdown in the browser they will all be enqueued around the same time so will be in the same queue that is being drained synchronously and since he's doing expensive calculations in all the callbacks it's going to freeze stuff.  I'd argue that callbacks being in different ticks is not part of the contract, they might be or they might not be, but for the specific use case switching to a browser crypto based verification of the signatures (which is async) would solve the problem.

source: I wrote the current version of the process.nextTick shim

**Comment by nolanlawson** [comment]
Looking at the [browserify shim](https://github.com/defunctzombie/node-process/blob/7d8c3702a8bbc43fa55f4bab74b150aef37001dd/browser.js#L110-L132), it seems to me that the problem is we're executing all items in the queue at once without taking into account how much time we've taken (and therefore how many frames we skipped). If we did a `performance.now()` and checked to see how long each task ran, then bail out and do another `setTimeout(fn, 0)` to process the remaining work after X number of milliseconds (let's say 10) it would fix @mikeal's issue while also potentially making a lot of other sites much faster.

Would this still be consistent with Node's implementation of `process.nextTick()`? If so, we could test it out on a few different web sites and see if a ~10ms maximum would be a good default.

**Comment by calvinmetcalf** [comment]
the node version of `process.nextTick` will do the node version of feezing the dom, and prevent events in the event loop from firing

**Comment by calvinmetcalf** [comment]
though thats only for recursive calls to it, not sure how it handles simultaneous calls

**Comment by nolanlawson** [comment]
> though thats only for recursive calls to it, not sure how it handles simultaneous calls 

Worth testing. I'd like to know too. :)

Either way though I'd argue that since the browser is a UI and Node is a server it might make sense to avoid blocking in the browser implementation even if it doesn't match 1-to-1 with the Node implementation. As long as it's not releasing Zalgo or otherwise causing interop issues, a maximum timeout of at most 16ms seems reasonable to me. The big downside is just overhead of `performance.now()` itself so it may need to be sampled rather than every turn of the loop.

The only alternative I can see is to tell folks to handle this themselves in userland and not invoke `process.nextTick()` too many times. AFAICT @mikeal's impl here just isolates the problem by creating a separate queue for memdown-only stuff as opposed to the global `nextTick` queue, which doesn't seem to solve that.

**Comment by calvinmetcalf** [comment]
you could also just use the setImmediate shim, the process.nextTick shim is designed to be lightweight and fulfill a similar role as process.nextTick in node

**Comment by vweevers** [comment]
@nolanlawson Any optimization like this will make userland optimization for a specific use case harder, esp. when there *is no* UI work to intersperse with (think Chrome extension, web worker, electron's main process). Or suppose I want to use `requestIdleCallback` and do as much work as possible until its deadline. I don't want 10ms timeouts there.

**Comment by nolanlawson** [comment]
@calvinmetcalf So what `setImmediate` does in Edge/IE is delay execution if there's an input or rendering task that's queued. It's unpolyfillable. Whereas there are lots of `setImmediate`/`immediate` shims for other browsers but my understanding is that they're either microtasks (i.e. input/render-blocking) or `setTimeout()` shims with a queue (i.e. same as the browserify `process.nextTick()`). So I'm not sure those would solve this issue either, although I'm happy to be corrected if I'm wrong on this. :smiley:

@vweevers I agree that's an issue; the 10ms maximum code could sniff to avoid running in web workers, etc. As for `requestIdleCallback()` if you're explicitly calling it, then I think that's a totally separate concern from a generic `nextTick()` impl. Also keep in mind I'm not suggesting `setTimeout(fn, 10)` I'm suggesting (`if (performance.now() - startTime > 10) { return setTimeout(processRestOfQueue, 0) }`

**Comment by calvinmetcalf** [comment]
The main setInmediate shim doesn't use a queue as of fairly recently

**Comment by nolanlawson** [comment]
I just tested, and Browserify 13.1.1 does not have a built-in shim for `setImmediate`. Closest I can find for  "main shim" is [the YuzuJS one](https://github.com/YuzuJS/setImmediate/blob/f1ccbfdf09cb93aadf77c4aa749ea554503b9234/setImmediate.js) which seems to prefer `postMessage()` in non-IE/non-Edge browsers.

I wasn't sure of the perf characteristics of `postMessage()`, so I put together [a small test](http://bl.ocks.org/nolanlawson/c04731d2d0ddf72acbcab84780ad21a6). According to my test, multiple `postMessage()`s queued simultaneously will block both input and rendering in Firefox 49 and Chrome 54, but don't block either (at least not completely) in Edge 14. So a `setImmediate()` shim would not help to reduce input/rendering blocking, in those browsers at least (need to test Safari still).

**Comment by calvinmetcalf** [comment]
I think it prefers mutation observer

**Comment by nolanlawson** [comment]
Looking at [the code](https://github.com/YuzuJS/setImmediate/blob/f1ccbfdf09cb93aadf77c4aa749ea554503b9234/setImmediate.js#L163-L182) it goes 1) Node nextTick, 2) postMessage(), 3) MessageChannel, 4) onreadystatechange w/ script 5) setTimeout.

**Comment by calvinmetcalf** [comment]
Hm maybe a pull there is in order

**Comment by nolanlawson** [comment]
There's no real polyfill, though. Either you wait a long time (setTimeout, requestIdleCallback) or you block input (postMessage, MutationObserver). Best solution I can think of is MutationObserver w/ queue and ~10ms limit. Or take this back to W3C and try to convince them setImmediate is cool. 😜

**Comment by calvinmetcalf** [comment]
ok so to circle around, I think part of the issue was just I felt like there were suggestions that this was a bug in process.nextTIck while from my perspective process.nextTick is working exactly as it is supposed to but what you want it s a [macrotask not a microtask](http://stackoverflow.com/a/25933985/1938759).  But I think @vweevers makes a good point about optimizations, I actually started to help maintain the process.nextTick shim after we realized the previous version, based on setTimeout without a queue, was adding too much latency to the tests of memdown.  

At the end of the day we can have low latency by default or we can have no render blocking by default, no render blocking is possible to build on top of low latency but the reverse is not possible.  So we should probably default to to low latency.  

Back to the original issue with @mikeal, it definitely is surprising that with this adapter, callbacks requested in the same tick, executing while blocking rendering, but I'm not sure how common his case is and the best solution all users just might be a note that multiple callbacks can be called without the event loop twirling and thus rendering can block if you request a bunch of items and do computationally expensive things in the callback and that something like using the promisified version of levelup or doing your computationally expensive thing async can prevent that.

**Comment by vweevers** [comment]
> At the end of the day we can have low latency by default or we can have no render blocking by default, no render blocking is possible to build on top of low latency but the reverse is not possible. So we should probably default to to low latency.

👍

**Comment by nolanlawson** [comment]
Good points all around. However a macrotask wouldn't help here because it would still lead to render-blocking. `for (var i = 0; i < 100; i++) { Promise.resolve().then((expensiveFunction); }` will still cause blocking because they'll all execute as a macrotask without yielding to anything.

> At the end of the day we can have low latency by default or we can have no render blocking by default

I believe we can have our cake and eat it too by opening a PR on `process.nextTick` to use the `performance.now()` measuring I outlined above. There's a sweet spot between yielding every callback to the browser via `setTimeout()` (which allows interleaving of input but has low latency) and running everything at once via a JS queue (which blocks rendering/input but has low latency). But obviously talk is cheap; I need write a benchmark and a PR to demonstrate it. :smiley:

In any case the current memdown impl (`global.setImmediate || process.nextTick`) is not ideal because it will have different perf characteristics in Edge/IE than in other browsers (namely it'll be high latency w/ zero blocking, whereas Chrome/FF/Safari will be low latency w/ high blocking). I'd propose sticking with just `process.nextTick` for now.

**Comment by calvinmetcalf** [comment]
> will still cause blocking because they'll all execute as a macrotask without yielding to anything.

macrotasks are supposed to yield after executing one, ie only one run per turn of the event loop

> I believe we can have our cake and eat it too by opening a PR on process.nextTick to use the performance.now() measuring I outlined above.

I somewhat doubt the other maintainers would go for that, the shim is widely included in a lot of places so its aim is to be lean as it could be to do it's aim which is to avoid zalgo by making things async, not to prevent render blocking.

That being said a macrotask library that uses performance.now might be very useful here, it just probably doesn't need to be by default included in webpack, browserify and a bunch of other strange edge cases and other bundlers.

**Comment by nolanlawson** [comment]
> macrotasks are supposed to yield after executing one, ie only one run per turn of the event loop

Sorry I misspoke; I read "microtask" 😜  You're right.

> it just probably doesn't need to be by default included in webpack, browserify and a bunch of other strange edge cases and other bundlers

I find this to be an interesting case where the default behavior in Node has wildly different consequences in the browser (since Node doesn't have to worry about blocking user input or rendering). There's definitely no universal right or wrong answer, but maybe Webpack/Browserify should have options to swap out different `process.nextTick` impl's. (E.g. an impl that just uses `setTimeout()` or uses `global.setImmediate || setTimeout` would avoid blocking at the cost of higher latency.)

**Comment by nolanlawson** [comment]
> no render blocking is possible to build on top of low latency but the reverse is not possible. So we should probably default to to low latency

BTW I'm not following this statement; I don't see how we can build a non-blocking solution on top of the current `process.nextTick` browser implementation. It just runs all tasks in the queue at once; a user of memdown wouldn't be able to break these up, although conceivably memdown itself could, although at a cost of a lot of extra code.

**Comment by calvinmetcalf** [comment]
The user could schedule their task to run later, they can't make it run
earlier.

I think I have a macrotask package I wrote a while back that we could
improve upon, it is currently "generational" where tasks of the same
generation are run without yielding, but any they enqueue are run in the
next event loop, we could totally build on this library.

And the thing about web pack is you can sub in whatever you want, including
bizzare choices like removing setTimeout which is why one must be very
careful with changes to process nextTick because nothing can be assumed.

**Comment by calvinmetcalf** [comment]
[found it](https://github.com/calvinmetcalf/macrotask)

**Comment by nolanlawson** [comment]
> The user could schedule their task to run later, they can't make it run earlier.

I suspect the overhead of the queue itself negates the benefits of deferring in a lot of cases (16ms is a very very low bar) but of course it always depends.

The macrotask library looks neat but yeah we'd need to build something on top of it, because looking at the code it appears to just run all first-generation tasks at once.

Without seeing @mikeal's test case though I think it's pretty much impossible to figure out what the best course of action is here. We've identified ~10 possible solutions but they may be possible to implement in userland rather than in memdown.

**Comment by nolanlawson** [comment]
So we've actually already discussed this in https://github.com/Level/memdown/pull/52 and the conclusion was to use `setImmediate` for Node but `require('immediate')` for the browser. I'm fine with this approach since it both acknowledges the need for different behavior in Node vs the browser and gives a consistent experience to all browsers, plus since it's low-latency/high-throughput as @calvinmetcalf points out you can still (hopefully) avoid render blocking by just wrapping your stuff in `setTimeout`.

I'll open a PR to do that; will close this issue for now unless we need more discussion

**Comment by calvinmetcalf** [comment]
> The macrotask library looks neat but yeah we'd need to build something on top of it, because looking at the code it appears to just run all first-generation tasks at once.

I'm working on a rewrite of it today
